### PR TITLE
bump libnetwork. vishvananda/netlink 1.0, vishvananda/netns

### LINKED ISSUE #136: dockerd --iptables=false adds DOCKER-USER chain and modify FORWARD chain anyway
### Issue description
* [x] This is a bug report

**dockerd** (in Docker version 17.09.0-ce, build afdb6d4) still adds the DOCKER-USER chain and modifies the FORWARD chain of the system, even when specifying the **--iptables=false** command line option.[ From the documentation](https://docs.docker.com/engine/reference/commandline/dockerd/#run-multiple-daemons):

> --iptables=false prevents the Docker daemon from adding iptables rules. 

### Actual behavior

1. **Initial system state**

```
# iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
```

2. **dockerd startup**

```
# dockerd                        --bridge=$BRIDGE_FOR_DOCKER                        \
                        --default-gateway=$VLAN_GW                     \
                        --dns=$VLAN_GW                                         \
                        --ip-forward=false                                         \
                        --ip-masq=false                                             \
                        --iptables=false                                              
```

3. **System state AFTER dockerd is started**

```
# iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination
DOCKER-USER  all  --  anywhere             anywhere

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain DOCKER-USER (1 references)
target     prot opt source               destination
RETURN     all  --  anywhere             anywhere
```

**Output of `docker version`:**

```
# docker --version
Docker version 17.09.0-ce, build afdb6d4

```

**Output of `dockerd`: with log-level=debug**

```
[...]
INFO[2017-10-18T19:29:00.127161968+02:00] Graph migration to content-addressability took 0.00 seconds
INFO[2017-10-18T19:29:00.128425657+02:00] Loading containers: start.
DEBU[2017-10-18T19:29:00.128541730+02:00] Option Experimental: false
DEBU[2017-10-18T19:29:00.128561767+02:00] Option DefaultDriver: bridge
DEBU[2017-10-18T19:29:00.128572980+02:00] Option DefaultNetwork: bridge
DEBU[2017-10-18T19:29:00.128601637+02:00] Network Control Plane MTU: 1500
DEBU[2017-10-18T19:29:00.137265625+02:00] Network (3d1968a) restored
DEBU[2017-10-18T19:29:00.147812823+02:00] Allocating IPv4 pools for network bridge (3d1968a69f6a809c876dd48f6863406bd98975bb0c766f56ca195a2d26e6c396)
DEBU[2017-10-18T19:29:00.147914464+02:00] RequestPool(LocalDefault, 10.0.110.0/24, , map[], false)
DEBU[2017-10-18T19:29:00.147981078+02:00] RequestAddress(LocalDefault/10.0.110.0/24, 10.0.110.7, map[RequestAddressType:com.docker.network.gateway])
DEBU[2017-10-18T19:29:00.148074662+02:00] RequestAddress(LocalDefault/10.0.110.0/24, 10.0.110.254, map[])
DEBU[2017-10-18T19:29:00.238350415+02:00] releasing IPv4 pools from network bridge (3d1968a69f6a809c876dd48f6863406bd98975bb0c766f56ca195a2d26e6c396)
DEBU[2017-10-18T19:29:00.238400391+02:00] ReleaseAddress(LocalDefault/10.0.110.0/24, 10.0.110.7)
DEBU[2017-10-18T19:29:00.238434285+02:00] ReleaseAddress(LocalDefault/10.0.110.0/24, 10.0.110.254)
DEBU[2017-10-18T19:29:00.238455487+02:00] ReleasePool(LocalDefault/10.0.110.0/24)
DEBU[2017-10-18T19:29:00.271778370+02:00] cleanupServiceBindings for 3d1968a69f6a809c876dd48f6863406bd98975bb0c766f56ca195a2d26e6c396
DEBU[2017-10-18T19:29:00.338727790+02:00] Allocating IPv4 pools for network bridge (b942ab7220a3a9d38c8edeb316d726f320a303c6fd51558e8e6ae068841642ce)
DEBU[2017-10-18T19:29:00.338785268+02:00] RequestPool(LocalDefault, 10.0.110.0/24, , map[], false)
DEBU[2017-10-18T19:29:00.338846284+02:00] RequestAddress(LocalDefault/10.0.110.0/24, 10.0.110.7, map[RequestAddressType:com.docker.network.gateway])
DEBU[2017-10-18T19:29:00.338894801+02:00] RequestAddress(LocalDefault/10.0.110.0/24, 10.0.110.254, map[])
DEBU[2017-10-18T19:29:00.443906556+02:00] Fail to initialize firewalld: Failed to connect to D-Bus system bus: dial unix /var/run/dbus/system_bus_socket: connect: no such file or directory, using raw iptables instead
DEBU[2017-10-18T19:29:00.447774069+02:00] /sbin/iptables, [--wait -t filter -n -L DOCKER-USER]
DEBU[2017-10-18T19:29:00.449089436+02:00] /sbin/iptables, [--wait -t filter -N DOCKER-USER]
DEBU[2017-10-18T19:29:00.450484274+02:00] /sbin/iptables, [--wait -t filter -C DOCKER-USER -j RETURN]
DEBU[2017-10-18T19:29:00.451875755+02:00] /sbin/iptables, [--wait -A DOCKER-USER -j RETURN]
DEBU[2017-10-18T19:29:00.453289482+02:00] /sbin/iptables, [--wait -t filter -C FORWARD -j DOCKER-USER]
DEBU[2017-10-18T19:29:00.454704161+02:00] /sbin/iptables, [--wait -I FORWARD -j DOCKER-USER]
INFO[2017-10-18T19:29:00.456151567+02:00] Loading containers: done.
INFO[2017-10-18T19:29:00.494739068+02:00] Docker daemon                                 commit=afdb6d4 graphdriver(s)=overlay2 version=17.09.0-ce
INFO[2017-10-18T19:29:00.494885547+02:00] Daemon has completed initialization
[...]
```

### LINKED ISSUE #2158: dockerd when run with --iptables=false modifies iptables by adding DOCKER-USER
If dockerd is run with --iptables=false option, it skips most of iptables stuff. But still interfere a bit. Here is what I get when run dockerd with --iptables=false:

```
*filter
:INPUT ACCEPT [12730:1832196]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [54:6656]
:DOCKER-USER - [0:0]
-A FORWARD -j DOCKER-USER
-A DOCKER-USER -j RETURN
```

Is it possible for docker not to touch iptables at all?

### LINKED ISSUE #21169: hostnames with new networking are case-sensitive
**Output of `docker version`:**

```
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   a34a1d5
 Built:        Fri Nov 20 17:56:04 UTC 2015
 OS/Arch:      darwin/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        2016-03-10T21:49:11.235199091+00:00
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 101
Images: 522
Server Version: 1.10.3
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 1205
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.19-boot2docker
Operating System: Boot2Docker 1.10.3 (TCL 6.4.1); master : 625117e - Thu Mar 10 22:09:02 UTC 2016
CPUs: 3
Total Memory: 1.463 GiB
Name: dev7
ID: C63I:2K5W:YZ5P:7M54:KD65:SVJE:QKDE:H5JS:BXHI:I27N:BOOP:GECU
Debug mode (server): true
 File Descriptors: 49
 Goroutines: 75
 System Time: 2016-03-13T21:06:17.199571395Z
 EventsListeners: 0
 Init SHA1: 
 Init Path: /usr/local/bin/docker
 Docker Root Dir: /mnt/sda1/var/lib/docker
Username: schmunk42
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

VirtualBox

**Steps to reproduce the issue:**
1. ping another host

**Describe the results you received:**

```
root@8eaa8bfbd08b:/app# ping HUB
ping: unknown host
root@8eaa8bfbd08b:/app# ping hub
PING hub (172.21.0.2): 56 data bytes
64 bytes from 172.21.0.2: icmp_seq=0 ttl=64 time=0.066 ms
```

**Describe the results you expected:**

An answer also for `ping HUB`

### LINKED ISSUE #28689: Embedded DNS is case-sensitive
Discovered this while discussing accepted characters for containers with @mstanleyjones;

```
docker network create foo
docker run -d --name FOO --net foo nginx:alpine
docker run -d --name foo --net foo nginx:alpine

docker run -it --rm --net foo alpine sh
/ # ping foo
PING foo (172.18.0.3): 56 data bytes
64 bytes from 172.18.0.3: seq=0 ttl=64 time=0.113 ms
64 bytes from 172.18.0.3: seq=1 ttl=64 time=0.099 ms
^C
--- foo ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.099/0.106/0.113 ms
/ # ping FOO
PING FOO (172.18.0.2): 56 data bytes
64 bytes from 172.18.0.2: seq=0 ttl=64 time=0.136 ms
64 bytes from 172.18.0.2: seq=1 ttl=64 time=0.096 ms
/ # ping fOo
ping: bad address 'fOo'
```

If looks like the embedded DNS is case sensitive, so resolving container IP addresses doesn't work if the wrong case is used. Also two containers on the same network can have the same name (but different case), which _should_ be a conflict, but currently isn't.

/cc @mavenugo

### LINKED ISSUE #35777: With iptables=false dockerd still creates DOCKER-USER chain and rules
**Description**

When `iptables=false`, dockerd still creates iptables chain and rules:

```text
-N DOCKER-USER
-A FORWARD -j DOCKER-USER
-A DOCKER-USER -j RETURN
```

**Steps to reproduce the issue:**
1. Fresh install of docker, set iptables=false in daemon.json or systemd unit file
2. systemctl start docker

**Describe the results you received:**

`iptables -S` produces after starting dockerd:

```text
-P INPUT ACCEPT
-P FORWARD ACCEPT
-P OUTPUT ACCEPT
-N DOCKER-USER
-A FORWARD -j DOCKER-USER
-A DOCKER-USER -j RETURN
```

**Describe the results you expected:**

`iptables -S` should produce:

```text
-P INPUT ACCEPT
-P FORWARD ACCEPT
-P OUTPUT ACCEPT
```

**Additional information you deem important (e.g. issue happens only occasionally):**

Repeatable.

**Output of `docker version`:**

```text
Client:
 Version:      17.09.1-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   19e2cf6
 Built:        Thu Dec  7 22:23:40 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.1-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   19e2cf6
 Built:        Thu Dec  7 22:25:03 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 17.09.1-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: journald
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-693.11.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 40
Total Memory: 62.71GiB
Name: fnl4b.bivio.biz
ID: PWPC:6VWA:6HIQ:K5MV:G2CK:TIHG:PSLM:HB73:DXXM:DQCI:2A4R:IFLC
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: true

WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

physical CentOS 7 box:

```
Linux fnl4b.bivio.biz 3.10.0-693.11.1.el7.x86_64 #1 SMP Mon Dec 4 23:52:40 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

### LINKED ISSUE #39184: Please update vishvananda/netns for ppc64 support
**Description**

The version of vishvananda/netns (vishvananda/netns@604eaf189ee867d8c147fafc28def2394e878d25) that is currenly vendored does not support ppc64 (only ppc64le).  ppc64 support was added by upstreams in commit vishvananda/netns@1fec6582c067519857603311ad090e23a5ca57de, over 3 years ago, so updating to this commit or any newer one fixes the issue.

**Steps to reproduce the issue:**
1. Try to build docker on linux/ppc64
2.
3.

**Describe the results you received:**
Compilation failed due to missing support in a vendored dependency

**Describe the results you expected:**
Compilation succeeded

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:           18.09.5
 API version:       1.39
 Go version:        go1.12.1
 Git commit:        e8ff056
 Built:             Wed May  1 16:42:46 2019
 OS/Arch:           linux/ppc64
 Experimental:      true

Server:
 Engine:
  Version:          18.09.5
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.12.1
  Git commit:       e8ff056
  Built:            Wed May  1 16:42:15 2019
  OS/Arch:          linux/ppc64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 2
Server Version: 18.09.5
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: bb71b10fd8f58240ca47fbb579b9d1028eea7c84
runc version: 2b18fe1d885ee5083ef9f0838fee39b62d653e30
init version: fec3683b971d9c3ef73f284f176672c44b448662
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.19.34
Operating System: Gentoo/Linux
OSType: linux
Architecture: ppc64
CPUs: 64
Total Memory: 63.68GiB
Name: hakua
ID: WJWV:AD7M:FA6K:IR3B:SJT7:RERM:PVPY:MDSL:6HX4:RXQV:JW26:3OU4
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: zeldin
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

### PR DESCRIPTION
**Pull request by thaJeztah**
full diffs:

- https://github.com/docker/libnetwork/compare/fc5a7d91d54cc98f64fc28f9e288b46a0bee756c...62a13ae87c6058bdc000fcccbf7e0f9f7525e2a9
- https://github.com/vishvananda/netlink/compare/b2de5d10e38ecce8607e6b438b6d174f389a004e...v1.0.0
- https://github.com/vishvananda/netns/compare/604eaf189ee867d8c147fafc28def2394e878d25...13995c7128ccc8e51e9a6bd2b551020a27180abd

notable changes in libnetwork:

- docker/libnetwork#2366 Bump vishvananda/netlink to 1.0.0
  - partially addresses https://github.com/moby/moby/issues/39404 Fork and update dependency of github.com/vishvananda/netns
  - fixes https://github.com/moby/moby/issues/39184 Please update vishvananda/netns for ppc64 support
- docker/libnetwork#2339 controller: Check if IPTables is enabled for arrangeUserFilterRule
  - fixes docker/libnetwork#2158 dockerd when run with --iptables=false modifies iptables by adding DOCKER-USER
  - fixes moby/moby#35777 With iptables=false dockerd still creates DOCKER-USER chain and rules
  - fixes docker/for-linux#136 dockerd --iptables=false adds DOCKER-USER chain and modify FORWARD chain anyway
- docker/libnetwork#2394 Make DNS records and queries case-insensitive
  - fixes moby/moby#28689 Embedded DNS is case-sensitive
  - fixes moby/moby#21169 hostnames with new networking are case-sensitive

### CODE REVIEW

**Comment by thaJeztah** [comment]
This is odd; it doesn't show merge-conflicts for me locally 🤔

File: hack/dockerfile/install/proxy.installer [hunk 1]
```diff
@@ -29,7 +29,7 @@ install_proxy_dynamic() {
 
 _install_proxy() {
 	echo "Install docker-proxy version $LIBNETWORK_COMMIT"
-	git clone https://github.com/docker/libnetwork.git "$GOPATH/src/github.com/docker/libnetwork"
+	git clone https://github.com/thaJeztah/libnetwork.git "$GOPATH/src/github.com/docker/libnetwork"
```
Comment (thaJeztah) [inline hack/dockerfile/install/proxy.installer]: WIP - using my fork

File: vendor.conf [hunk 2]
```diff
@@ -39,7 +39,7 @@ github.com/gofrs/flock 7f43ea2e6a643ad441fc12d0ecc0d3388b300c53 # v0.7.0
 #get libnetwork packages
 
 # When updating, also update LIBNETWORK_COMMIT in hack/dockerfile/install/proxy.installer accordingly
-github.com/docker/libnetwork 48f846327bbe6a0dce0c556e8dc9f5bb939d5c16
+github.com/docker/libnetwork 7b661439fd0d143e8d50ef57294f41a4a4e50e1a https://github.com/thaJeztah/libnetwork.git
```
Comment (thaJeztah) [inline vendor.conf]: WIP - using my fork

**Comment by thaJeztah** [comment]
ping @euanh @arkodg PTAL

**Review by vdemeester — APPROVED** [review]
LGTM :tiger:

**Review by tiborvass — APPROVED** [review]

File: vendor.conf [hunk 3]
```diff
@@ -50,8 +50,8 @@ github.com/hashicorp/go-sockaddr                    6d291a969b86c4b633730bfc6b8b
 github.com/hashicorp/go-multierror                  fcdddc395df1ddf4247c69bd436e84cfa0733f7e
 github.com/hashicorp/serf                           598c54895cc5a7b1a24a398d635e8c0ea0959870
 github.com/docker/libkv                             458977154600b9f23984d9f4b82e79570b5ae12b
-github.com/vishvananda/netns                        604eaf189ee867d8c147fafc28def2394e878d25
-github.com/vishvananda/netlink                      b2de5d10e38ecce8607e6b438b6d174f389a004e
+github.com/vishvananda/netns                        13995c7128ccc8e51e9a6bd2b551020a27180abd
+github.com/vishvananda/netlink                      a2ad57a690f3caf3015351d2d6e1c0b95c349752 # v1.0.0
```
Comment (tiborvass) [inline vendor.conf]: @thaJeztah FYI I just realized that this is not v1.0.0. Maybe we want to keep that in mind for cherry-picking.

Comment (thaJeztah) [reply vendor.conf]: I just double checked,  but looks like this commit matches the v1.0.0 tag? https://github.com/vishvananda/netlink/commit/a2ad57a690f3caf3015351d2d6e1c0b95c349752

Comment (tiborvass) [reply vendor.conf]: omg... they have a v1.0.0 branch that's not the tag...

Comment (tiborvass) [reply vendor.conf]: opened https://github.com/vishvananda/netlink/issues/465

Comment (thaJeztah) [reply vendor.conf]: oooooooh!

I wonder which one `vndr` picks if we don't specify the sha 🤔

**Pull request merged**